### PR TITLE
dwallet-helios: add `contract_interactions` functions & tests

### LINKED ISSUE #9: dwallet-helios: add contract_interactions module.

### PR DESCRIPTION
**Pull request by yuvalelmo**
This PR includes functions that are not in use **yet**.
The functions are in use in a future PR - #6.

This PR is created in order to make this code more reviewable (shorter PRs).

### CODE REVIEW

File: execution/src/utils.rs
Comment (ycscaly) [inline execution/src/utils.rs:1]: Please explain the purpose of this file, and where the code originates.
It looks copied-pasted, if so from where?
If not, what specification did you follow in implementation?

Comment (ycscaly) [reply execution/src/utils.rs:1]: Please add a module level documentation explaining the purpose and need of these utilities

**Review by ycscaly — CHANGES_REQUESTED** [review]

File: execution/src/utils.rs
Comment (yuvalelmo) [reply execution/src/utils.rs:1]: It is not copy-pasted from anywhere, but I'll take that as a compliment 💯 
I've added the documentation as requested.

File: execution/src/utils.rs [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+/// Solidity's documentation for storage slots and mapping layouts.
+///
+/// Solidity uses 256-bit hashes (H256) for storage slots and keys.
+/// When converting integers or other types of data to fit this format, padding is required
+/// to ensure they match the expected 256-bit length.
+///
+/// The code in this module originates from the need to interact with smart contract storage
+/// layouts programmatically, ensuring consistency and correctness in storage access patterns.
+/// The implementation follows the specifications outlined in the Solidity documentation:
+/// https://docs.soliditylang.org/en/v0.8.24/internals/layout_in_storage.html#mappings-and-dynamic-arrays
```
Comment (ycscaly) [inline execution/src/utils.rs]: Great!

Comment (ycscaly) [reply execution/src/utils.rs]: Please convert to module-level documentation: https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html

File: execution/src/utils.rs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+/// This function standardizes the input slot for a given unsigned 64-bit integer.
+/// It first converts the integer into a hexadecimal string representation.
+/// Then, it pads the hexadecimal string to ensure it has a length of 64 characters.
+/// We pad the string because in solidity, the slot is a 256-bit hash (H256).
+/// Finally,
+/// it decodes the padded hexadecimal string back into bytes and converts it into a 256-bit hash
+/// (H256).
```
Comment (ycscaly) [inline execution/src/utils.rs]: is this a standard you designed or was this a part of the specification?

File: execution/src/utils.rs
Comment (ycscaly) [inline execution/src/utils.rs:1]: why is this file and logic apart of the fork, and not of the dwallet-network repo?

**Review by ycscaly — CHANGES_REQUESTED** [review]

File: execution/src/utils.rs (hunk 2 above)
Comment (yuvalelmo) [reply execution/src/utils.rs]: This standard is part of the Solidity specifications

File: execution/src/utils.rs
Comment (yuvalelmo) [reply execution/src/utils.rs:1]: because they contain logic that is used in this repo (i.e. for fetching proofs)

Comment (ycscaly) [reply execution/src/utils.rs:1]: at least in this PR I don't see any proof being fetched based on this logic

Comment (yuvalelmo) [reply execution/src/utils.rs:1]: you are correct, one of these functions (get_message_storage_slot) is used here:
https://github.com/dwallet-labs/dwallet-helios/pull/6

the rest of the functions are being used only by this function so i changed them to be private now.

File: execution/src/utils.rs [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+//! Solidity's documentation for storage slots and mapping layouts.
+//!
+//! Solidity uses 256-bit hashes (H256) for storage slots and keys.
+//! When converting integers or other types of data to fit this format, padding is required
+//! to ensure they match the expected 256-bit length.
+//!
+//! The code in this module originates from the need to interact with smart contract storage
+//! layouts programmatically, ensuring consistency and correctness in storage access patterns.
+//! The implementation follows the specifications outlined in the Solidity documentation:
+//! https://docs.soliditylang.org/en/v0.8.24/internals/layout_in_storage.html#mappings-and-dynamic-arrays
```
Comment (zeevmoney) [inline execution/src/utils.rs]:
```suggestion
//! <https://docs.soliditylang.org/en/v0.8.24/internals/layout_in_storage.html#mappings-and-dynamic-arrays>
```

always use `cargo doc` command to check the build

File: execution/src/lib.rs [hunk 4]
```diff
@@ -9,3 +9,4 @@ mod execution;
 pub use crate::execution::*;
 
 mod proof;
+mod utils;
```
Comment (zeevmoney) [inline execution/src/lib.rs]:
```suggestion
pub mod utils;
```

File: execution/src/utils.rs
Comment (zeevmoney) [inline execution/src/utils.rs:1]: rename the file to`contract_interactions.rs`

File: execution/src/utils.rs [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+use ethers::types::H256;
+use eyre::Error;
+use sha3::{Digest, Keccak256};
+
+/// This function standardizes the input slot for a given unsigned 64-bit integer.
```
Comment (zeevmoney) [inline execution/src/utils.rs]:
```suggestion
/// Standardizes the input slot for a given unsigned 64-bit integer.
```

File: execution/src/utils.rs [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+
+/// This function standardizes the input key for a given 256-bit hash (H256).
+/// It first converts the hash into a hexadecimal string representation.
+/// Then, it pads the hexadecimal string to ensure it has a length of 64 characters.
+/// We pad the string because in solidity, the slot is a 256-bit hash (H256).
+/// Finally,
+/// it decodes the padded hexadecimal string back into bytes and converts it into a 256-bit hash
+/// (H256).
+/// # Arguments
+/// * `input` - A 256-bit hash (H256) that represents the input key.
```
Comment (zeevmoney) [inline execution/src/utils.rs]:
```suggestion
/// * `input` – A 256-bit hash (H256) that represents the input key.
```

File: execution/src/utils.rs [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+/// First initializes a new `Keccak256` hasher, then standardizes the input slot and key.
+/// The standardized key and slot are then hashed together to produce a new `H256` hash.
+/// The result hash will be used to get the location of the
+/// (key, value) pair in the contract's storage.
+/// # Arguments
+/// * `key` - A H256 hash that represents the key for which the mapping slot is to be calculated.
+///   The Key is `Keccak256(message + dwallet_id)`.
+/// * `Mapping_slot` - A `u64` value that represents the mapping slot in the contract storage layout.
+///   For more info:
+///   https://docs.soliditylang.org/en/v0.8.24/internals/layout_in_storage.html#mappings-and-dynamic-arrays
```
Comment (zeevmoney) [inline execution/src/utils.rs]:
```suggestion
/// * `key` – A H256 hash that represents the key for which the mapping slot is to be calculated.
///   The Key is `Keccak256(message + dwallet_id)`.
/// * `Mapping_slot` – A `u64` value that represents the mapping slot in the contract storage layout.
///   [For more info](https://docs.soliditylang.org/en/v0.8.24/internals/layout_in_storage.html#mappings-and-dynamic-arrays)
```

Comment (zeevmoney) [reply execution/src/utils.rs]: Regular links in docs should have <http://foo.bar>, or you can create a clickable link test with 
\[test\]\(https://foo.bar\)

File: execution/src/utils.rs [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    let mut hasher = Keccak256::new();
+    let slot_padded = standardize_slot_input(mapping_slot);
+    let key_padded = standardize_key_input(key);
+    hasher.update(key_padded.as_bytes());
+    hasher.update(slot_padded.as_bytes());
+    H256::from_slice(&hasher.finalize())
+}
+
+/// Calculates the key for a given message and dWallet ID.
+/// In the smart contract, the key is calculated by hashing the message and the dWallet id together.
```
Comment (zeevmoney) [inline execution/src/utils.rs]:
```suggestion
/// In the smart contract, the key is calculated by hashing the message, and the dWallet ID together.
```
Words in names that are initialisms or acronyms (e.g., "URL" or "NATO") have a consistent case.

https://github.com/dwallet-labs/wiki/wiki/Rust-Development-Standards#:~:text=Words%20in%20names%20that%20are%20initialisms%20or%20acronyms%20(e.g.%2C%20%22URL%22%20or%20%22NATO%22)%20have%20a%20consistent%20case.

File: execution/src/utils.rs [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+/// Then, it calculates the mapping slot for the calculated key and the provided data slot.
+/// The calculated mapping slot can be used to locate the (key, value) pair in the contract's
+/// storage. # Arguments
+/// * `message` - A string that represents the message to be stored.
+/// * `dwallet_id` - A vector of bytes that represents the dWallet ID.
+/// * `data_slot` - An unsigned 64-bit integer that represents the data slot.
+/// # Returns
+/// * A `Result` that contains a 256-bit hash (H256) that represents the calculated storage slot, or
+///   an `Error` if the calculation fails.
+
```
Comment (zeevmoney) [inline execution/src/utils.rs]:
```suggestion
/// Calculates the storage slot for a given message, dWallet ID, and data slot.
/// The function first calculates a key by hashing the message, and the dWallet ID together.
/// Then, it calculates the mapping slot for the calculated key, and the provided data slot.
/// The calculated mapping slot can be used to locate the (key, value) pair in the contract's
/// storage. # Arguments
/// * `message` – A string that represents the message to be stored.
/// * `dwallet_id` – A vector of bytes that represents the dWallet ID.
/// * `data_slot` – An unsigned 64-bit integer that represents the data slot.
/// # Returns
/// * A `Result` that contains a 256-bit hash (H256) that represents the calculated storage slot, or
///   an `Error` if the calculation fails.
```

**Review by zeevmoney — CHANGES_REQUESTED** [review]
see comments

**Review by zeevmoney — APPROVED** [review]
gj

**Review by ycscaly — CHANGES_REQUESTED** [review]
Again, I am against implementing Solidity standards, and I mentioned that this might be much more difficult than it is in this PR.

We need to properly consider how to handle this

**Review by ycscaly — APPROVED** [review]
I did not review that the code implements the specification correctly. We should have an external review on this or dedicate more time to learn the Solidity specification

**Comment by zeevmoney** [comment]
> I did not review that the code implements the specification correctly. We should have an external review on this or dedicate more time to learn the Solidity specification

I checked ethers code, they basically do the same thing, and @yuvalelmo 's code is correct according to the spec link that's in the doc.
If the code is not correct, it just won't work (since it's based on a hash, and if the hash is not correct, the key in the mapping won't be found).

**Pull request merged**